### PR TITLE
Refactor Transform String Resugar Term Scott 

### PR DESCRIPTION
**Pull request by Liberxue**
#713 
* Create  struct to wrap user-provided cons handling functions, implementing the  trait.
* Implement  struct to handle default cons patterns, also implementing the  trait.
* Replace  with  type, better expressing success or failure semantics.
* Utilize Rust's pattern matching and  method to simplify error handling process.

### CODE REVIEW

**Review by developedby — CHANGES_REQUESTED** [review]
The visitor is completely unnecessary here, doesn't really simplify.

I like the idea of the `try_resugar` functions, it lets the pattern matching not be repeated twice.

You could have a single `build_string` function that takes takes the encoding, tries the correct encoded pattern and tries the non-applied constructors, without repetition.

Also, please leave the `if` tree since it mimics the shape of the AST and is easier for me to visualize the structure.

**Comment by Liberxue** [comment]
> The visitor is completely unnecessary here, doesn't really simplify.
> 
> I like the idea of the `try_resugar` functions, it lets the pattern matching not be repeated twice.
> 
> You could have a single `build_string` function that takes takes the encoding, tries the correct encoded pattern and tries the non-applied constructors, without repetition.
> 
> Also, please leave the `if` tree since it mimics the shape of the AST and is easier for me to visualize the structure.

Done Please review～～thx @developedby

File: src/fun/transform/resugar_string.rs [hunk 1]
```diff
@@ -7,101 +7,132 @@ impl Term {
...
+
+        // Try custom cons handler
+        if let Some((head, tail)) = cons_handler(current) {
+          s.push(head);
+          current = tail;
+          continue;
+        }
+
+        // Cons: @x (x CONS_TAG <num> <str>)
+        if let Term::Lam { tag: Tag::Static, pat, bod } = current {
```
Comment (developedby) [inline src/fun/transform/resugar_string.rs:123]: This is specifically the "num-scott" encoding, you shouldn't apply it for all cases

**Review by developedby — CHANGES_REQUESTED** [review]
This is nicer to read without the visitor, but my other previous comments still apply.

Also, please do the changes on `resugar_list` as well. It's almost an exact copy of this one, just looking for a slightly different pattern. It would be confusing if they were implemented differently

**Comment by Liberxue** [comment]
> This is nicer to read without the visitor, but my other previous comments still apply.
> 
> Also, please do the changes on `resugar_list` as well. It's almost an exact copy of this one, just looking for a slightly different pattern. It would be confusing if they were implemented differently

In the `resugar_lists` file, after careful reading, I found that extracting code blocks could lead to reduced performance due to cloning. I haven't thought of a better approach at the moment~~~

**Review by developedby — APPROVED** [review]

File: src/fun/transform/resugar_list.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@ impl Term {
   /// Converts scott-encoded lists ending with List/Nil to list literals.
   fn resugar_lists_scott(&mut self) {
     maybe_grow(|| {
-      // Search for a List/Cons pattern in the term and try to build a list from that point on.
+      // Searc      h for a List/Cons pattern in the term and try to build a list from that point on.
```
Comment (developedby) [inline src/fun/transform/resugar_list.rs:82]:
```suggestion
      // Search for a List/Cons pattern in the term and try to build a list from that point on.
```

**Comment by developedby** [comment]
Can you just run `cargo fmt`?

**Comment by Liberxue** [comment]
> Can you just run `cargo fmt`?
So Sorry OKK~~ done

**Comment by developedby** [comment]
Thank you very much for the contribution

**Pull request merged**